libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/route/PrimaryRouteUpdateTask.java

```diff
@@ -1,66 +1,97 @@
 package com.mapbox.services.android.navigation.ui.v5.route;
 
-import android.os.AsyncTask;
+import android.os.Handler;
 
 import com.mapbox.geojson.Feature;
 import com.mapbox.geojson.FeatureCollection;
 
 import java.lang.ref.WeakReference;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 import static com.mapbox.services.android.navigation.ui.v5.route.RouteConstants.FIRST_COLLECTION_INDEX;
 import static com.mapbox.services.android.navigation.ui.v5.route.RouteConstants.PRIMARY_ROUTE_PROPERTY_KEY;
 
-class PrimaryRouteUpdateTask extends AsyncTask<Void, Void, List<FeatureCollection>> {
+class PrimaryRouteUpdateTask extends Thread {
 
   private final int newPrimaryIndex;
   private final List<FeatureCollection> routeFeatureCollections;
   private final WeakReference<OnPrimaryRouteUpdatedCallback> callbackWeakReference;
+  private AtomicBoolean cancelThread = new AtomicBoolean(false);
+  private Handler mainThreadHandler;
 
   PrimaryRouteUpdateTask(int newPrimaryIndex, List<FeatureCollection> routeFeatureCollections,
-                         OnPrimaryRouteUpdatedCallback callback) {
+                         OnPrimaryRouteUpdatedCallback callback, Handler handler) {
     this.newPrimaryIndex = newPrimaryIndex;
     this.routeFeatureCollections = routeFeatureCollections;
     this.callbackWeakReference = new WeakReference<>(callback);
+    this.mainThreadHandler = handler;
```

Comment:
we don't know about handler, it could be non-Main.
```suggestion
    this.postHandler = handler;
```